ca_tools/scripts/pos_controller_pid.py

```diff
@@ -0,0 +1,212 @@
+
+    def get_time(self):
+        """
+        Function to get the current time.
+        """
+
+        self._now = rospy.get_time()
+        if self._last_time < 0.001:
+            self._last_time = self._now
+        self._time_change = self._now - self._last_time
+
+    def update_time(self):
+        """
+        Function to update the Last Time.
+        """
+
+        if self._now > 0:
+            self._last_time = self._now
+
+    def set_goals(self):
+        """
+        Function that sets the new goals checking the state and if the _set_variable flag
+        is turned True, if it is true, means that a new goal needs to be set.
+        """
+
+
+        if self._state == "FORWARD" and self._set_variable:
+            self._input_angle = self._current_angle
+            self._ini_aux.x = self._curr_aux.x
+            self._ini_aux.y = self._curr_aux.y
+            self._fin_aux.x = self._ini_aux.x + 2 * math.cos(self._input_angle)
+            self._fin_aux.y = self._ini_aux.y + 2 * math.sin(self._input_angle)
+            self._set_length = self._fin_aux
+            self._set_angle = self._input_angle
+            self._set_variable = False
+
+        if self._state == "TURN" and self._set_variable:
+            self._input_length = self._ini_aux
+            self._input_angle = self._current_angle
+            self._set_length = self._input_length
+            self._set_angle = self._input_angle + math.pi / 2
+            self._set_angle = math.atan2(math.sin(self._set_angle), math.cos(self._set_angle))
+            self._set_variable = False
+
+    def pid(self):
+        """
+        Function that calculates the error and publish the velocity constantly.
+        """
+
+        # Error calculations
+
+        self._error_angle = self._set_angle - self._current_angle
+        self._err_sum_angle += self._error_angle * self._time_change
+        self._is_new_goal.x = self._set_length.x - self._curr_aux.x
+        self._is_new_goal.y = self._set_length.y - self._curr_aux.y
+        self._error_length = math.hypot(self._is_new_goal.x, self._is_new_goal.y)
+        self._err_sum_length += self._err_sum_length * self._time_change
+
+        if self._state == "FORWARD":
+            self._output_length = self._kp * self._error_length + self._ki * self._err_sum_length
+            self._output_angle = 0.0
+
+        if self._state == "TURN":
+            self._output_angle = self._kp * self._error_angle + self._ki * self._err_sum_angle
+            self._output_length = 0.0
+
+        if self._output_length > self._max_output_length:
+            self._output_length = self._max_output_length
+        if self._output_length < 0:
+            self._output_length = self._max_output_length
+        if self._output_angle > self._max_output_angle or self._output_angle < 0:
+            self._output_angle = self._max_output_angle
+
+        self._vel.linear.x = self._output_length
+        self._vel.angular.z = self._output_angle
```

Comment:
Beware this is an angular speed measured in `rad/s` and you are assigning an angle.
It is not completely wrong, but you should be careful that the values are not the same. Specially in the case of angular speed, it doesn't make sense to clip an angle, because the robot can turn as fast as it goes.